File: rust/src/publisher.rs
/// Pub/sub dispatcher + WebSocket server for feed handler consumers.
///
/// Fans out BBO and book updates from feed handler output channels to
/// multiple subscribers, with optional per-symbol and per-stream filtering.
use std::collections::HashSet;
use std::sync::Arc;

use axum::extract::ws::{Message, WebSocket};
use axum::extract::WebSocketUpgrade;
use axum::response::IntoResponse;
use axum::routing::get;
use log::info;
use serde_json::Value;
use tokio::sync::{mpsc, Mutex};

use crate::feed_handler::{BboEvent, BookEvent};

// === CONFIGURATION ===

const SUBSCRIBER_QUEUE_MAX: usize = 10_000;

// === TYPES ===

#[derive(Debug, Clone, PartialEq)]
enum StreamFilter {
    All,
    Bbo,
    Book,
}

struct Subscriber {
    tx: mpsc::Sender<Value>,
    stream: StreamFilter,
    symbols: Option<HashSet<String>>,
}

// === STATE ===

static SUBSCRIBERS: std::sync::LazyLock<Arc<Mutex<Vec<Subscriber>>>> =
    std::sync::LazyLock::new(|| Arc::new(Mutex::new(Vec::new())));

// === PUB/SUB ===

pub async fn subscribe(
    stream: String,
    symbols: Option<Vec<String>>,
) -> mpsc::Receiver<Value> {
    let (tx, rx) = mpsc::channel(SUBSCRIBER_QUEUE_MAX);
    let filter = match stream.as_str() {
        "bbo" => StreamFilter::Bbo,
        "book" => StreamFilter::Book,
        _ => StreamFilter::All,
    };
    let sym_set = symbols.map(|s| s.into_iter().collect::<HashSet<String>>());
    let mut subs = SUBSCRIBERS.lock().await;
    subs.push(Subscriber {
        tx,
        stream: filter,
        symbols: sym_set,
    });
    info!("Subscriber added ({} total)", subs.len());
    rx
}

pub async fn unsubscribe(target_tx: &mpsc::Sender<Value>) {
    let mut subs = SUBSCRIBERS.lock().await;
    subs.retain(|s| !s.tx.same_channel(target_tx));
    info!("Subscriber removed ({} remaining)", subs.len());
}

fn put_value(tx: &mpsc::Sender<Value>, msg: Value) {
    let _ = tx.try_send(msg);
}

async fn publish(msg: Value, stream: &StreamFilter) {
    let symbol = msg.get("symbol").and_then(|v| v.as_str()).unwrap_or("");
    let subs = SUBSCRIBERS.lock().await;
    for sub in subs.iter() {
        if sub.stream != *stream && sub.stream != StreamFilter::All {
            continue;
        }
        if let Some(ref symbols) = sub.symbols {
            if !symbols.contains(symbol) {
                continue;
            }
        }
        put_value(&sub.tx, msg.clone());
    }
}

// === DISPATCHER ===

pub async fn run_dispatcher(
    mut bbo_rx: mpsc::Receiver<BboEvent>,
    mut book_rx: mpsc::Receiver<BookEvent>,
) {
    let bbo_handle = tokio::spawn(async move {
        while let Some(event) = bbo_rx.recv().await {
            let msg = serde_json::to_value(&event).unwrap_or_default();
            publish(msg, &StreamFilter::Bbo).await;
        }
    });

    let book_handle = tokio::spawn(async move {
        while let Some(event) = book_rx.recv().await {
            let msg = serde_json::to_value(&event).unwrap_or_default();
            publish(msg, &StreamFilter::Book).await;
        }
    });

    let _ = tokio::join!(bbo_handle, book_handle);
}

// === WEBSOCKET SERVER ===

async fn ws_handler(ws: WebSocketUpgrade) -> impl IntoResponse {
    ws.on_upgrade(handle_ws_client)
}

async fn handle_ws_client(mut socket: WebSocket) {
    info!("WebSocket client connected");
    let mut sub_tx: Option<mpsc::Sender<Value>> = None;
    let mut stream_handle: Option<tokio::task::JoinHandle<()>> = None;

    loop {
        tokio::select! {
            msg = socket.recv() => {
                match msg {
                    Some(Ok(Message::Text(text))) => {
                        let parsed: Value = match serde_json::from_str(&text) {
                            Ok(v) => v,
                            Err(_) => {
                                let _ = socket.send(Message::Text(r#"{"error":"invalid JSON"}"#.into())).await;
                                continue;
                            }
                        };

                        let action = parsed.get("action").and_then(|v| v.as_str()).unwrap_or("");

                        if action == "subscribe" {
                            if let Some(h) = stream_handle.take() { h.abort(); }
                            if let Some(ref tx) = sub_tx {
                                unsubscribe(tx).await;
                            }

                            let stream = parsed.get("stream").and_then(|v| v.as_str()).unwrap_or("all").to_string();
                            let symbols: Option<Vec<String>> = parsed.get("symbols").and_then(|v| {
                                v.as_array().map(|a| a.iter().filter_map(|s| s.as_str().map(String::from)).collect())
                            });

                            let mut rx = subscribe(stream.clone(), symbols.clone()).await;
                            sub_tx = None;

                            let confirmation = serde_json::json!({
                                "status": "subscribed",
                                "stream": stream,
                                "symbols": symbols,
                            });
                            let _ = socket.send(Message::Text(confirmation.to_string().into())).await;

                            let (out_tx, mut out_rx) = mpsc::channel::<String>(1000);
                            stream_handle = Some(tokio::spawn(async move {
                                while let Some(val) = rx.recv().await {
                                    let json_str = serde_json::to_string(&val).unwrap_or_default();
                                    if out_tx.send(json_str).await.is_err() {
                                        break;
                                    }
                                }
                            }));

                            let mut streaming = true;
                            while streaming {
                                tokio::select! {
                                    Some(out_msg) = out_rx.recv() => {
                                        if socket.send(Message::Text(out_msg.into())).await.is_err() {
                                            streaming = false;
                                        }
                                    }
                                    msg = socket.recv() => {
                                        match msg {
                                            Some(Ok(Message::Text(t))) => {
                                                if let Ok(v) = serde_json::from_str::<Value>(&t) {
                                                    if v.get("action").and_then(|a| a.as_str()) == Some("unsubscribe") {
                                                        if let Some(h) = stream_handle.take() { h.abort(); }
                                                        let _ = socket.send(Message::Text(r#"{"status":"unsubscribed"}"#.into())).await;
                                                        streaming = false;
                                                    }
                                                }
                                            }
                                            Some(Ok(Message::Close(_))) | None => {
                                                streaming = false;
                                            }
                                            _ => {}
                                        }
                                    }
                                }
                            }
                        }
                    }
                    Some(Ok(Message::Close(_))) | None => break,
                    _ => {}
                }
            }
        }
    }

    if let Some(h) = stream_handle.take() {
        h.abort();
    }
    info!("WebSocket client disconnected");
}

pub async fn start_ws_server(port: u16) -> anyhow::Result<()> {
    let app = axum::Router::new().route("/ws", get(ws_handler));
    let listener = tokio::net::TcpListener::bind(format!("0.0.0.0:{}", port)).await?;
    info!("WebSocket server on :{}/ws", port);
    tokio::spawn(async move {
        axum::serve(listener, app).await.ok();
    });
    Ok(())
}

File: python/publisher.py
"""Pub/sub dispatcher + WebSocket server for feed handler consumers.

Fans out BBO and book updates from feed_handler's output queues to
multiple subscribers, with optional per-symbol and per-stream filtering.

Internal consumers: call subscribe() for a filtered asyncio.Queue.
External consumers: connect via WebSocket, send a subscribe message.

WebSocket protocol:
    -> {"action": "subscribe", "stream": "bbo", "symbols": ["BTCUSDT", "ETHUSDT"]}
    <- {"status": "subscribed", "stream": "bbo", "symbols": ["BTCUSDT", "ETHUSDT"]}
    <- {"symbol": "BTCUSDT", "bid_price": "64000.50", "ask_price": "64001.00", ...}
    -> {"action": "unsubscribe"}
    <- {"status": "unsubscribed"}
"""
import asyncio
import json
import logging
from decimal import Decimal

from aiohttp import web

logger = logging.getLogger(__name__)

# === CONFIGURATION ===

SUBSCRIBER_QUEUE_MAX = 10_000

# === STATE ===

_subscribers: list[dict] = []

# Messages shed because a subscriber's queue was full (slow consumer).
# Never silent: surfaced in /stats and /metrics.
_dropped = 0


def dropped_count() -> int:
    return _dropped


def subscriber_count() -> int:
    return len(_subscribers)


# === SERIALIZATION ===

def _default(o):
    if isinstance(o, Decimal):
        return str(o)
    raise TypeError(f"not serializable: {type(o)}")


# === PUB/SUB ===

def subscribe(stream: str = "all", symbols: set[str] | None = None) -> asyncio.Queue:
    """Register a consumer. Returns a queue that receives matching updates.

    stream: "bbo", "book", or "all"
    symbols: set of symbol names to filter, or None for all symbols
    """
    q: asyncio.Queue = asyncio.Queue(maxsize=SUBSCRIBER_QUEUE_MAX)
    _subscribers.append({
        "queue": q,
        "symbols": set(symbols) if symbols else None,
        "stream": stream,
    })
    logger.info("Subscriber added: stream=%s, symbols=%s (%d total)",
                stream, "all" if symbols is None else len(symbols), len(_subscribers))
    return q


def unsubscribe(q: asyncio.Queue):
    """Remove a subscriber by its queue reference."""
    _subscribers[:] = [s for s in _subscribers if s["queue"] is not q]
    logger.info("Subscriber removed (%d remaining)", len(_subscribers))


# === DISPATCHER ===

def _put(q: asyncio.Queue, msg: dict):
    """Enqueue, shedding the OLDEST message on overflow. Counted, never silent.

    Cross-language divergence, deliberate and documented: Rust's tokio mpsc has
    no sender-side pop, so the Rust publisher sheds the NEWEST message instead.
    Both count into the same metric name.
    """
    global _dropped
    try:
        q.put_nowait(msg)
        return
    except asyncio.QueueFull:
        pass
    _dropped += 1
    try:
        q.get_nowait()
    except asyncio.QueueEmpty:
        pass
    try:
        q.put_nowait(msg)
    except asyncio.QueueFull:
        pass


def _publish(msg: dict, stream: str):
    symbol = msg.get("symbol")
    for sub in _subscribers:
        if sub["stream"] not in ("all", stream):
            continue
        if sub["symbols"] is not None and symbol not in sub["symbols"]:
            continue
        _put(sub["queue"], msg)


async def run_dispatcher(bbo_q: asyncio.Queue, book_q: asyncio.Queue):
    """Drain feed handler queues and fan out to all subscribers."""
    async def drain(source_q: asyncio.Queue, stream_name: str):
        while True:
            msg = await source_q.get()
            _publish(msg, stream_name)

    await asyncio.gather(
        drain(bbo_q, "bbo"),
        drain(book_q, "book"),
    )


# === WEBSOCKET SERVER ===

async def _stream_to_ws(ws: web.WebSocketResponse, q: asyncio.Queue):
    try:
        while not ws.closed:
            msg = await q.get()
            await ws.send_str(json.dumps(msg, default=_default))
    except (ConnectionResetError, asyncio.CancelledError):
        pass


async def _ws_handler(request: web.Request) -> web.WebSocketResponse:
    ws = web.WebSocketResponse()
    await ws.prepare(request)
    logger.info("WebSocket client connected from %s", request.remote)

    q = None
    stream_task = None

    try:
        async for raw in ws:
            if raw.type != web.WSMsgType.TEXT:
                continue
            try:
                msg = json.loads(raw.data)
            except json.JSONDecodeError:
                await ws.send_str('{"error": "invalid JSON"}')
                continue

            action = msg.get("action")

            if action == "subscribe":
                if stream_task is not None:
                    stream_task.cancel()
                if q is not None:
                    unsubscribe(q)

                stream = msg.get("stream", "all")
                symbols = msg.get("symbols")
                q = subscribe(stream=stream, symbols=symbols)
                stream_task = asyncio.create_task(_stream_to_ws(ws, q))
                await ws.send_str(json.dumps({
                    "status": "subscribed",
                    "stream": stream,
                    "symbols": symbols,
                }))

            elif action == "unsubscribe":
                if stream_task is not None:
                    stream_task.cancel()
                    stream_task = None
                if q is not None:
                    unsubscribe(q)
                    q = None
                await ws.send_str('{"status": "unsubscribed"}')

    finally:
        if stream_task is not None:
            stream_task.cancel()
        if q is not None:
            unsubscribe(q)
        logger.info("WebSocket client disconnected from %s", request.remote)

    return ws


async def start_ws_server(port: int = 8081):
    app = web.Application()
    app.router.add_get("/ws", _ws_handler)
    runner = web.AppRunner(app)
    await runner.setup()
    await web.TCPSite(runner, "0.0.0.0", port).start()
    logger.info("WebSocket server on :%d/ws", port)
